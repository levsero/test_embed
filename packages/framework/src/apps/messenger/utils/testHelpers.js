import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
import ThemeProvider from 'src/apps/messenger/features/themeProvider'
import createStore from 'src/apps/messenger/store'
import hostPageWindow from 'src/framework/utils/hostPageWindow'

export const render = (ui, { render, store, themeProps = {} } = {}) => {
  const reduxStore = store || createStore()

  const renderFn = render || rtlRender
  return {
    store: reduxStore,
    ...renderFn(
      <Provider store={reduxStore}>
        <ThemeProvider theme={themeProps}>{ui}</ThemeProvider>
      </Provider>
    )
  }
}

export const mockMatchMedia = () => {
  // Keep track of all media queries made and their callbacks
  // {
  //   <media query>: [ <callback1>, <callback2> ]
  // }
  const mediaQueries = {}

  const mockMatchMedia = jest.fn().mockImplementation(query => {
    if (!mediaQueries[query]) {
      mediaQueries[query] = []
    }

    return {
      matches: true,
      addEventListener: (_, callback) => {
        mediaQueries[query].push(callback)
      }
    }
  })

  const triggerChangeForBreakpoint = (breakpoint, event) => {
    mediaQueries[breakpoint]?.forEach(callback => callback(event))
  }

  hostPageWindow.matchMedia = mockMatchMedia

  return {
    triggerChangeForBreakpoint
  }
}

/*
  For testing reducers (duh).
  Pass it your reducer and an array of parameter objects.
  - If parameter object is a flat action payload, the test will use
  snapshot testing. The action can also be explicitly specified using
  an `action` key.
  - If parameter object contains key `initialState`, the value
  of that key will be used as initial state for the reducer.
  - If parameter object contains key `expected`, the value
  of that key will be used to assert against the result of the reducer.
*/
export const testReducer = (reducer, actions) => {
  actions.forEach(params => {
    const { expected, initialState, extraDesc } = params
    const action = params.action || params
    const basicTestDesc = `${reducer.name}, action: ${action.type}`
    const testDesc = extraDesc ? `${basicTestDesc} ${extraDesc}` : basicTestDesc

    test(testDesc, () => {
      const reduced = reducer(initialState, action)

      if ('expected' in params) {
        expect(reduced).toEqual(expected)
      } else {
        expect(reduced).toMatchSnapshot()
      }
    })
  })
}
