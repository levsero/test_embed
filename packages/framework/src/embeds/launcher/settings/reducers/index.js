import { combineReducers } from 'redux'

import label from 'src/embeds/launcher/settings/reducers/label'
import talkLabel from 'src/embeds/launcher/settings/reducers/talkLabel'
import chatLabel from 'src/embeds/launcher/settings/reducers/chatLabel'

export default combineReducers({
  label,
  chatLabel,
  talkLabel
})
