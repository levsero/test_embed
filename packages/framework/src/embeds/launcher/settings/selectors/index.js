import { createSelector } from 'reselect'
import { i18n } from 'src/apps/webWidget/services/i18n'
import { getTalkTitleKey } from 'src/embeds/talk/selectors'
import { getLocale } from 'src/redux/modules/base/base-selectors'

const getLabel = (_, label) => label

export const getSettingsLauncherLabel = (state) => state.settings.launcher.settings.label
export const getSettingsLauncherChatLabel = (state) => state.settings.launcher.settings.chatLabel
export const getSettingsLauncherTalkLabel = (state) => state.settings.launcher.settings.talkLabel

export const getLauncherTalkLabel = createSelector(
  [getSettingsLauncherTalkLabel, getTalkTitleKey, getLocale],
  (settingsLauncherTalkLabel, talkTitleKey, _locale) =>
    i18n.getSettingTranslation(settingsLauncherTalkLabel) || i18n.t(talkTitleKey)
)

export const getLauncherChatLabel = createSelector(
  [getSettingsLauncherChatLabel, getLocale],
  (settingsLauncherChatLabel, _locale) =>
    i18n.getSettingTranslation(settingsLauncherChatLabel) ||
    i18n.t('embeddable_framework.launcher.label.chat')
)

export const getLauncherLabel = createSelector(
  [getSettingsLauncherLabel, getLocale, getLabel],
  (settingsLauncherLabel, _locale, label) =>
    i18n.getSettingTranslation(settingsLauncherLabel) || i18n.t(label)
)
