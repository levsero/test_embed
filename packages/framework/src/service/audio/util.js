import { includes } from 'lodash'

const canPlayTypeResultWhitelist = ['maybe', 'probably']

export function getSupportedAudioType(audio) {
  if (!audio.canPlayType) {
    return
  }

  const canPlayType = type => includes(canPlayTypeResultWhitelist, audio.canPlayType(type))

  if (canPlayType('audio/mpeg')) {
    return 'mp3'
  } else if (canPlayType('audio/ogg; codecs="vorbis"')) {
    return 'ogg'
  } else if (canPlayType('audio/wav')) {
    return 'wav'
  }
}
