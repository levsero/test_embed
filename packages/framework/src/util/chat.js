import _ from 'lodash'
import { AGENT_BOT } from 'constants/chat'
import { i18n } from 'src/apps/webWidget/services/i18n'
import { win, getZendeskHost } from 'utility/globals'

/**
 * Return a new array with consecutive numbers combined into an array containing only the start and end (inclusive) number.
 * @param {NumberLike[]} numbers
 * @returns {NumberRange[]} numberRanges
 */
function combineNumbers(numbers) {
  const result = []
  let last, start

  const dump = () => (start = (result.push(start ? [start, last] : last), undefined))

  numbers.forEach((num) => {
    num = parseInt(num, 10)
    if (last === num - 1) {
      if (!start) start = last
    } else if (last) dump()
    last = num
  })

  dump()
  return result
}

const periodsToString = (periods) =>
  periods.map((period) => `${period.start},${period.end}`).toString()
const stringToPeriods = (str) =>
  str
    ? _.chunk(str.split(','), 2).map((arr) => ({
        start: parseInt(arr[0], 10),
        end: parseInt(arr[1], 10),
      }))
    : []

/**
 * Return a FormattedSchedule array that is more amendable to the Widget's rendering needs.
 * @param {Schedule} schedule
 * @return {FormattedSchedule[]}
 */
function formatSchedule(schedule) {
  schedule = { ...schedule, 7: schedule[0] } // use 7 for Sunday
  delete schedule[0]
  let preprocessedDates = _.invertBy(_.mapValues(schedule, periodsToString))

  return _.sortBy(
    _.map(preprocessedDates, (days, periods) => ({
      periods: stringToPeriods(periods),
      days: combineNumbers(days),
    })),
    ['days']
  )
}

function isAgent(nick) {
  return nick.indexOf('agent:') > -1 && nick !== AGENT_BOT
}

function isVisitor(nick) {
  return nick.indexOf('visitor') > -1 && nick !== AGENT_BOT
}

function isDefaultNickname(name) {
  const nameRegex = new RegExp(/^Visitor [0-9]{3,}$/)

  return nameRegex.test(name)
}

function getDisplayName(inName, fallbackName = '') {
  return isDefaultNickname(inName) ? fallbackName : inName
}

function createChatPopoutWindow(chatPopoutSettings, machineId, locale) {
  let url

  if (__DEV__) {
    url = '/liveChat.html'
  } else {
    url = `${__ASSET_BASE_PATH__}/web_widget/latest/liveChat.html`
  }

  url += generateQueryString(chatPopoutSettings, machineId, locale)

  win.open(url, 'WebWidgetLiveChat', 'height=600,width=400')
}

function encodeSettings(settings) {
  return win.btoa(encodeURIComponent(JSON.stringify(settings)))
}

function generateQueryString(settings, machineId, locale) {
  const subdomain = getZendeskHost(document)

  const encodedSettings = encodeSettings(settings)

  const title = encodeURI(i18n.t('embeddable_framework.popout.title'))

  return `?v=10#key=${subdomain}&settings=${encodedSettings}&mid=${machineId}&locale=${locale}&title=${title}`
}

function cleanBrandName(brand) {
  if (brand) {
    return brand.replace(/,/g, '')
  }
}

export {
  combineNumbers,
  formatSchedule,
  isDefaultNickname,
  isAgent,
  isVisitor,
  createChatPopoutWindow,
  getDisplayName,
  cleanBrandName,
}

/**
 * A number, or a string containing a number.
 * @typedef {(number|string)} NumberLike
 */

/**
 * A number representing a single number, or a 2-element array representing a range of consecutive numbers.
 * @typedef {(number|[number, number])} NumberRange
 */

/**
 * An object representing an operating period. Values refer to the number of minutes since midnight
 * (0 means midnight, 1439 means 23:59)
 * @typedef {object} Period
 * @property {number} start
 * @property {number} end
 */

/**
 * An object containing an operating hours schedule. 0 is for Sunday, 1 for Monday, etc.
 * The properties do NOT actually have underscore prefixed.
 * @typedef {object} Schedule
 * @property {Period[]} _0
 * @property {Period[]} _1
 * @property {Period[]} _2
 * @property {Period[]} _3
 * @property {Period[]} _4
 * @property {Period[]} _5
 * @property {Period[]} _6
 */

/**
 * An object containing operating periods and the applicable days. Sunday is represented by 7 here.
 * @typedef {object} FormattedSchedule
 * @property {NumberRange[]} days
 * @property {Periods[]} periods
 */
