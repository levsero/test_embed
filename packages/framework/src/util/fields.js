import React from 'react'
import _ from 'lodash'
import sanitizeHtml from 'sanitize-html'

import { i18n } from 'src/apps/webWidget/services/i18n'

const setupConditionCheck = (customFields, formState) => {
  return (fieldId, value) => {
    const field = _.find(customFields, (field) => field.id === fieldId)

    if (!field) return false

    if (field.type === 'checkbox') {
      // classic wants 0 and 1 so we use those as the values but conditions give us true and false
      return value === !!formState[fieldId]
    }

    return value === formState[fieldId]
  }
}

const getConditionOverrides = (conditions, conditionCheck) =>
  _.reduce(
    conditions,
    (memo, condition) => {
      const isFulfilled = conditionCheck(condition.parent_field_id, condition.value)

      condition.child_fields.forEach((child) => {
        // need to check if already set to false in case multiple conditions on the same element
        const isVisible = memo[child.id]
          ? memo[child.id].visible_in_portal || isFulfilled
          : isFulfilled
        const isRequired =
          (memo[child.id] ? memo[child.id].required_in_portal : false) ||
          (child.is_required && isFulfilled)

        memo[child.id] = {
          visible_in_portal: isVisible, // eslint-disable-line camelcase
          required_in_portal: isRequired, // eslint-disable-line camelcase
        }
      })
      return memo
    },
    {}
  )

export const updateConditionalVisibility = (customFields, formState, conditions) => {
  const conditionCheck = setupConditionCheck(customFields, formState)
  const conditionOverrides = getConditionOverrides(conditions, conditionCheck)

  const fields = _.map(customFields, (field) => {
    return { ...field, ...conditionOverrides[field.id] }
  })

  return fields
}

const shouldRenderErrorMessage = (value, required, showErrors, pattern) => {
  const isRequiredCheckValid = !required || value
  const isPatternCheckValid = !value || (pattern ? pattern.test(value) : true)
  const isValid = isRequiredCheckValid && isPatternCheckValid

  return !isValid && showErrors
}

const getStyledLabelText = (label, required) => {
  if (!label) return null

  // Disallow all HTML elements in label argument, it will render as pure text
  const sanitizedLabel = sanitizeHtml(label, { allowedTags: [] })

  return required
    ? `<strong>${sanitizedLabel}</strong>`
    : i18n.t('embeddable_framework.validation.label.new_optional', {
        label: sanitizedLabel,
      })
}

const renderLabel = (Component, label, required) => {
  const labelText = getStyledLabelText(label, required)

  return <Component dangerouslySetInnerHTML={{ __html: labelText }} />
}

export { shouldRenderErrorMessage, renderLabel, getStyledLabelText }
