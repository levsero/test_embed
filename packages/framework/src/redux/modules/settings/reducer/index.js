import { combineReducers } from 'redux'

import answerBot from './answerBot'
import chat from './chat'
import analytics from './analytics'
import cookies from './cookies'
import launcher from './launcher'
import color from './color'
import helpCenter from './help-center'
import styling from './styling'
import contactForm from './contactForm'
import talk from './talk'
import contactOptions from './contactOptions'
import navigation from './navigation'

export default combineReducers({
  analytics,
  answerBot,
  chat,
  color,
  cookies,
  contactForm,
  contactOptions,
  helpCenter,
  launcher,
  navigation,
  styling,
  talk
})
