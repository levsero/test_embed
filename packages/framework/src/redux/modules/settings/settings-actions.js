import _ from 'lodash'
import { CONNECTION_STATUSES } from 'src/constants/chat'
import { setDepartment } from 'src/redux/modules/chat'
import { getConnection, getZChatVendor } from 'src/embeds/chat/selectors'
import { getDefaultSelectedDepartment } from 'src/redux/modules/selectors'
import { getSettingsChatTags } from 'src/redux/modules/settings/settings-selectors'
import { settings as legacySettings } from 'src/service/settings'
import { UPDATE_SETTINGS } from './settings-action-types'

const handleTagsChange = (zChat, tags = [], oldTags = []) => {
  if (_.isEqual(tags, oldTags) || !_.isArray(tags) || !_.isArray(oldTags)) {
    return
  }

  const toAdd = _.difference(tags, oldTags)
  const toRemove = _.difference(oldTags, tags)

  toAdd.length > 0 && zChat.addTags(toAdd)
  toRemove.length > 0 && zChat.removeTags(toRemove)
}

export function updateSettings(settings) {
  return (dispatch, getState) => {
    if (!_.get(settings, 'webWidget')) {
      settings = {
        webWidget: {
          ...settings,
        },
      }
    }

    if (settings?.webWidget?.authenticate?.chat?.jwtFn) {
      legacySettings.storeChatAuth(settings.webWidget.authenticate.chat.jwtFn)
    }

    if (settings?.webWidget?.authenticate?.jwtFn) {
      legacySettings.storeHelpCenterAuth(settings.webWidget.authenticate.jwtFn)
    }

    const oldTags = getSettingsChatTags(getState())

    dispatch({
      type: UPDATE_SETTINGS,
      payload: settings,
    })

    const state = getState()

    if (getConnection(state) === CONNECTION_STATUSES.CONNECTED) {
      dispatch(updateChatSettings(oldTags))
    }
  }
}

export function updateChatSettings(oldTags) {
  return (dispatch, getState) => {
    const state = getState()
    const tags = getSettingsChatTags(state)
    const zChat = getZChatVendor(state)
    const visitorDepartmentId = _.get(getDefaultSelectedDepartment(state), 'id')

    handleTagsChange(zChat, tags, oldTags)

    if (visitorDepartmentId) {
      dispatch(setDepartment(visitorDepartmentId))
    }
  }
}
