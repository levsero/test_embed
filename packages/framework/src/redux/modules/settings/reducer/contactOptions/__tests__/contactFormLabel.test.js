import { UPDATE_SETTINGS } from 'src/redux/modules/settings/settings-action-types'
import { testReducer } from 'src/util/testHelpers'
import contactFormLabel from '../contactFormLabel'

const badPayload = {
  foo: 'bar',
}
const goodPayload = {
  webWidget: {
    contactOptions: {
      contactFormLabel: { '*': 'form label' },
    },
  },
}

testReducer(contactFormLabel, [
  { type: undefined, payload: '' },
  { type: 'DERP DERP', payload: '' },
  { type: UPDATE_SETTINGS, payload: badPayload },
  { type: UPDATE_SETTINGS, payload: goodPayload },
])
