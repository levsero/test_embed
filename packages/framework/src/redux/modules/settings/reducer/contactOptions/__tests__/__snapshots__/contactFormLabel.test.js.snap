// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contactFormLabel, action: DERP DERP 1`] = `null`;

exports[`contactFormLabel, action: undefined 1`] = `null`;

exports[`contactFormLabel, action: widget/settings/UPDATE_SETTINGS 1`] = `null`;

exports[`contactFormLabel, action: widget/settings/UPDATE_SETTINGS 2`] = `
Object {
  "*": "form label",
}
`;
